src/bare_item.rs

```diff
@@ -0,0 +1,410 @@
+    /// ```
+    /// # use sfv::BareItem;
+    /// let bare_item = BareItem::ByteSeq("foo".to_owned().into_bytes().into());
+    /// assert_eq!(bare_item.as_byte_seq().unwrap().as_slice(), "foo".as_bytes());
+    /// ```
+    pub fn as_byte_seq(&self) -> Option<&Vec<u8>> {
+        match *self {
+            BareItem::ByteSeq(ref val) => Some(&val.0),
+            _ => None,
+        }
+    }
+    /// If `BareItem` is a `Boolean`, returns `bool`, otherwise returns `None`.
+    /// ```
+    /// # use sfv::{BareItem, Decimal, FromPrimitive};
+    /// let bare_item = BareItem::Boolean(true.into());
+    /// assert_eq!(bare_item.as_bool().unwrap(), true);
+    /// ```
+    pub fn as_bool(&self) -> Option<bool> {
+        match self {
+            BareItem::Boolean(val) => Some(val.0),
+            _ => None,
+        }
+    }
+    /// If `BareItem` is a `Token`, returns `&str`, otherwise returns `None`.
+    /// ```
+    /// use sfv::BareItem;
+    /// # use std::convert::TryInto;
+    /// # fn main() -> Result<(), &'static str> {
+    ///
+    /// let bare_item = BareItem::Token("*bar".try_into()?);
+    /// assert_eq!(bare_item.as_token().unwrap(), "*bar");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn as_token(&self) -> Option<&str> {
+        match *self {
+            BareItem::Token(ref val) => Some(val),
+            _ => None,
+        }
+    }
+}
+
+impl TryFrom<i64> for BareItem {
+    type Error = &'static str;
+    /// Converts `i64` into `BareItem::Integer`.
+    /// ```
+    /// # use sfv::BareItem;
+    /// # use std::convert::TryInto;
+    /// # fn main() -> Result<(), &'static str> {
+    /// let bare_item: BareItem = 456_i64.try_into()?;
+    /// assert_eq!(bare_item.as_int().unwrap(), 456);
+    /// # Ok(())
+    /// # }
+    /// ```
+    fn try_from(item: i64) -> Result<Self, Self::Error> {
+        Ok(BareItem::Integer(item.try_into()?))
+    }
+}
+
+impl TryFrom<rust_decimal::Decimal> for BareItem {
+    type Error = &'static str;
+    /// Converts `rust_decimal::Decimal` into `BareItem::Decimal`.
+    /// ```
+    /// # use sfv::{BareItem, FromPrimitive};
+    /// # use std::convert::TryInto;
+    /// use rust_decimal::Decimal;
+    /// # fn main() -> Result<(), &'static str> {
+    /// let decimal_number = Decimal::from_f64(48.01).unwrap();
+    /// let bare_item: BareItem = decimal_number.try_into()?;
+    /// assert_eq!(bare_item.as_decimal().unwrap(), decimal_number);
+    /// # Ok(())
+    /// # }
+    /// ```
+    fn try_from(item: rust_decimal::Decimal) -> Result<Self, Self::Error> {
+        Ok(BareItem::Decimal(item.try_into()?))
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Decimal(pub(crate) rust_decimal::Decimal);
+
+impl TryFrom<rust_decimal::Decimal> for Decimal {
+    type Error = &'static str;
+    fn try_from(value: rust_decimal::Decimal) -> Result<Self, Self::Error> {
+        let mut output = String::new();
+        Serializer::serialize_decimal(value, &mut output)?;
```

Comment:
So, doing an actual serialization in the try_from implementation seems less than optimal.
That means for every bare item that we construct, we allocate a string we then throw away. And then we do the serialization again when we actually serialize it into a buffer.

I think what we should do here is split the serialization from the validity checks, so try_from for Decimal would only check if the decimal is in the allowed range, and not actually serialize it.

Reply:
Yeah I'm not happy with validating by serializing either, and as mentioned in the PR description we could (and should 😅) move validations to a place where both can access it.
To get my idea validated and not change another hundreds of lines, I decided to keep it like this for the moment - but I'm totally supportive of the fact that this needs to change.

Reply:
Created a new `ValidateValue` trait that the value structs implement. It will either return a `Resul` where the `Ok` value is either the input (passing back ownership) or a new "sanitized" value (important for `Decimal` as we can round to three decimal places already)